File: src/sphere.cpp
#include "./sphere.h"

bool Sphere::hit(const Ray& r, double t_min, double t_max, HitRecord& rec) const
{
    auto oc = r.origin() - _center;
    auto a = dot(r.direction(), r.direction());
    auto b = dot(oc, r.direction());
    auto c = dot(oc, oc) - _radius * _radius;
    auto discriminant = b * b - a * c;

    if (discriminant < 0.0)
        return false;

    double temp = (-b - sqrt(discriminant)) / a;
    if (temp > t_min  && temp < t_max)
    {
        rec.t = temp;
        rec.p = r.point_at_parameter(temp);
        rec.normal = (rec.p - _center) / _radius;
        rec.material = _material;
        return true;
    }

    temp = (-b + sqrt(discriminant)) / a;
    if (temp > t_min  && temp < t_max)
    {
        rec.t = temp;
        rec.p = r.point_at_parameter(temp);
        rec.normal = (rec.p - _center) / _radius;
        rec.material = _material;
        return true;
    }

    return false;
}

File: rust/src/sphere.rs
use crate::hittable::Hittable;
use crate::hittable::HitRecord;
use crate::ray::Ray;
use crate::vector::Point;

#[derive(Debug, Copy, Clone)]
pub struct Sphere {
    pub center: Point,
    pub radius: f64,
}

impl Hittable for Sphere {
    fn hit(&self, ray: &Ray, t_min: f64, t_max: f64) -> Result<HitRecord, bool> {
        let oc = ray.orig - self.center;
        let a = ray.dir.length_squared();
        let half_b = oc.dot(&ray.dir);
        let c = oc.length_squared() - self.radius * self.radius;
        
        let discriminant = half_b * half_b - a * c;
        if discriminant < 0.0 {
            return Err(false);
        }

        let sqrtd = discriminant.sqrt();
        let mut root = (-half_b - sqrtd) / a;
        if root < t_min || t_max < root {
            root = (-half_b + sqrtd) / a;
            if root < t_min || t_max < root {
                return Err(false);
            }
        }

        let impact = ray.at(root);
        let mut rec = HitRecord {
            point: impact,
            normal: (impact - self.center) / self.radius,
            t: root,
            front_face: true,
        };
        let outward_normal = (rec.point - self.center) / self.radius;
        rec.set_face_normal(ray, outward_normal);

        return Ok(rec);
    }
}
